universal-application-tool-0.0.1/app/services/applicant/Block.java

```diff
@@ -56,11 +56,58 @@ public String getDescription() {
   }
 
   public boolean hasErrors() {
-    if (errorsMemo.isEmpty()) {
-      this.errorsMemo =
-          getQuestions().stream().map(ApplicantQuestion::hasErrors).reduce(Boolean::logicalOr);
+    return getQuestions().stream()
+        .map(ApplicantQuestion::hasErrors)
+        .reduce(Boolean::logicalOr)
+        .orElse(false);
+  }
+
+  /**
+   * Checks whether the block is complete - that is, {@link ApplicantData} has values at all the
+   * paths for all required questions in this block and there are no errors. Note: this cannot be
+   * memoized, since we need to reflect internal changes to ApplicantData.
+   */
+  public boolean isCompleteWithoutErrors() {
+    // TODO(https://github.com/seattle-uat/civiform/issues/551): Stream only required scalar paths
+    // instead of all scalar paths.
+    return blockDefinition.scalarPaths().stream()
+            .filter(path -> !applicantData.hasValueAtPath(path))
+            .findAny()
+            .isEmpty()
+        && !hasErrors();
+  }
+
+  /**
+   * Checks whether all questions in this block were completed while applying to the given program.
+   *
+   * @param programId the program ID to check
+   * @return true if all questions were last updated while filling out the program with the given
+   *     ID; false otherwise
+   */
+  public boolean wasCompletedInProgram(long programId) {
+    return getQuestions().stream()
+        .allMatch(
```

Comment:
Suggestion (blocking): if any of the questions were answered in this program it should be included in this program, otherwise a block could disappear after an applicant completes it?

```suggestion
        .anyMatch(
```

Reply:
Thanks! Updated